### PR TITLE
Feature: Added api for getting/setting the kv_cache

### PR DESCRIPTION
**Pull request by chrfalch**
This PR contains a simple extension to the C-API for getting/setting the kv_cache so that an app can save the state of the kv_cache after providing a prompt and load this next time the app starts to avoid having to evaluate the prompt on startup.

The api provides access methods for retrieving the current memory buffer for the kv_cache and its token number. It also contains a method for setting the kv_cache from a memory buffer and a token count.

I did not start implementing the `--cache-prompt` argument since it is a bit more involved - we need to save some more state like the last_n_tokens and n_past parameter. We'd also need to hash the prompt, check if a prompt file existed etc.

Implements foundation for #64

### CODE REVIEW

**Comment by chrfalch** [comment]
Example app using the new api to load the state is able to answer with previous knowledge from the saved prompt (uses the chat example):

image

File: llama.cpp [hunk 1]
```diff
@@ -1668,6 +1668,33 @@ int llama_model_quantize(
     return 0;
 }
 
+// Returns the KV cache that will contain the context for the
+// ongoing prediction with the model.
+uint8_t* llama_get_kv_cache(struct llama_context * ctx) {
```
Comment (prusnak) [inline llama.cpp:1673]:
```suggestion
const uint8_t * llama_get_kv_cache(struct llama_context * ctx) {
```

File: llama.cpp [hunk 2]
```diff
@@ -1668,6 +1668,33 @@ int llama_model_quantize(
...
+}
+
+int llama_get_kv_cache_token_count(struct llama_context * ctx) {
+  return ctx->model.kv_self.n;
+}
+
+// Sets the KV cache containing the current context for the model
+void llama_set_kv_cache(
+        struct llama_context * ctx,
+                     uint8_t * kv_cache,
```
Comment (prusnak) [inline llama.cpp:1689]:
```suggestion
                     const uint8_t * kv_cache,
```

File: llama.h [hunk 3]
```diff
@@ -83,6 +83,23 @@ extern "C" {
             const char * fname_out,
                    int   itype);
 
+    // Returns the KV cache that will contain the context for the
+    // ongoing prediction with the model.
+    LLAMA_API uint8_t* llama_get_kv_cache(struct llama_context * ctx);
```
Comment (prusnak) [inline llama.h:88]:
```suggestion
    LLAMA_API const uint8_t * llama_get_kv_cache(struct llama_context * ctx);
```

File: llama.h [hunk 4]
```diff
@@ -83,6 +83,23 @@ extern "C" {
...
+    // Returns the size of the KV cache
+    LLAMA_API size_t llama_get_kv_cache_size(struct llama_context * ctx);
+
+    // Returns the number of tokens in the KV cache
+    LLAMA_API int llama_get_kv_cache_token_count(struct llama_context * ctx);
+
+    // Sets the KV cache containing the current context for the model
+    LLAMA_API void llama_set_kv_cache(
+            struct llama_context * ctx,
+                         uint8_t * kv_cache,
```
Comment (prusnak) [inline llama.h:99]:
```suggestion
                         const uint8_t * kv_cache,
```

**Review by prusnak — CHANGES_REQUESTED** [review]
use `const` where appropriate

**Comment by chrfalch** [comment]
Thanks for the comments, @prusnak - appreciate it.

**Review by ggerganov — APPROVED** [review]
Very good start for solving #64 

Use 4 spaces indentation

**Review by prusnak — APPROVED** [review]

**Pull request merged**